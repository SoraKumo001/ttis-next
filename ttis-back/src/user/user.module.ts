import { Module } from '@nestjs/common';
import { UserResolver } from './user.resolver';
import { TypeOrmModule } from '@nestjs/typeorm';
import { User } from './user';

@Module({
  imports: [TypeOrmModule.forFeature([User])],
  providers: [UserResolver],
})
export class UserModule {}
