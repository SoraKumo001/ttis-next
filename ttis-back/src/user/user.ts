import { Entity, PrimaryGeneratedColumn, Column, BaseEntity } from 'typeorm';
import { ObjectType, Field, Int } from '@nestjs/graphql';

@ObjectType()
@Entity()
export class User extends BaseEntity {
  @Field(_ => Int)
  @PrimaryGeneratedColumn() //自動番号
  id!: number;
  @Field()
  @Column({ default: true })
  enable!: boolean;
  @Field()
  @Column({ unique: true })
  name!: string;
  @Column({ nullable: true })
  password?: string;
  @Field()
  @Column({ default: '{}' })
  info!: string;
}
