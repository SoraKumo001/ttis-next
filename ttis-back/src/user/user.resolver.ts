import { Resolver, Args, Mutation, Query, Info, Int } from '@nestjs/graphql';
import { User } from './user';
import { getFields } from '@libs/graphQLTools';
import { UseGuards } from '@nestjs/common';
import { JwtAuthGuard, CurrentUser } from '../auth/auth.guard';
import { UserService } from './user.service';

@Resolver('User')
export class UserResolver {
  constructor(private readonly service: UserService) {}

  @UseGuards(JwtAuthGuard)
  @Query((_) => [User], { nullable: true })
  async users(@CurrentUser() user, @Info() info) {
    if (!user) return null;
    return this.service.users(getFields(info));
  }

  @UseGuards(JwtAuthGuard)
  @Mutation((_) => User, { nullable: true })
  async createUser(
    @Args('name') name: string,
    @Args('password') password: string,
    @Args('info', { nullable: true }) info?: string,
    @CurrentUser() user?,
  ) {
    if (!user) return null;
    return this.service.setUser(
      undefined,
      name,
      password,
      info && JSON.parse(info),
    );
  }

  @UseGuards(JwtAuthGuard)
  @Mutation((_) => User, { nullable: true })
  async updateUser(
    @Args('id', { type: () => Int }) id: number,
    @Args('name', { nullable: true }) name?: string,
    @Args('password', { nullable: true }) password?: string,
    @Args('info', { nullable: true }) info?: string,
    @CurrentUser() user?,
  ) {
    if (!user) return null;
    return this.service.setUser(id, name, password, info && JSON.parse(info));
  }

  @UseGuards(JwtAuthGuard)
  @Mutation((_) => Boolean)
  async deleteUser(
    @CurrentUser() user,
    @Args('id', { type: () => Int }) id: number,
  ) {
    if (!user) return false;
    return this.service.deleteUser(id);
  }
  @UseGuards(JwtAuthGuard)
  @Mutation((_) => Boolean)
  async deleteUsers(
    @Args('ids', { type: () => [Int] }) ids: number[],
    @CurrentUser() user?,
  ) {
    if (!user) return false;
    return this.service.deleteUsers(ids);
  }

}
