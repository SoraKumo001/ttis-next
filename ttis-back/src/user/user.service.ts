import { Injectable, OnModuleInit } from '@nestjs/common';
import { User } from './user';
import { Repository } from 'typeorm';
import { InjectRepository } from '@nestjs/typeorm';
import { GraphQLFeald } from '@libs/graphQLTools';
import { getSHA256 } from '@libs/hashTools';

@Injectable()
export class UserService implements OnModuleInit {
  constructor(
    @InjectRepository(User)
    private readonly rep: Repository<User>,
  ) {}
  async onModuleInit() {
    const { rep } = this;
    return rep.manager.connection.transaction(async () => {
      return rep.count().then((count) => {
        if (!count) {
          return rep.save({ name: 'admin', password: getSHA256('') });
        }
      });
    });
  }
  async setUser(id?: number, name?: string, password?: string, info?: unknown) {
    const { rep } = this;
    if (!id) {
      if (name && password) {
        return rep.save({
          name,
          password: getSHA256(password),
          info: info && JSON.stringify(info),
        });
      }
    } else {
      const user = await rep.findOne(id);
      if (user) {
        if (name) user.name = name;
        if (password) user.password = getSHA256(password);
        return user.save();
      }
    }
    return null;
  }
  async users(fields?: GraphQLFeald) {
    const { rep } = this;
    return rep.find({
      select: fields as (keyof User)[],
      where: { enable: true },
      order: { id: 'ASC' },
    });
  }
  async deleteUser(id: number) {
    const { rep } = this;
    return !!(await rep.update(id, { enable: false }));
  }

  async deleteUsers(ids: number[]) {
    const { rep } = this;
    return !!(await rep.update(ids, { enable: false }));
  }
  async login(name: string, password: string) {
    const { rep } = this;
    const passwordHash = getSHA256(password);
    return await rep.findOne({
      where: { name, password: passwordHash, enable: true },
    });
  }
  async validateUser(id: number, password: string) {
    const { rep } = this;
    return await rep.findOne({
      where: { id, password, enable: true },
    });
  }
}
