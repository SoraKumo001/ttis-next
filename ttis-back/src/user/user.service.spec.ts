import { Test, TestingModule } from '@nestjs/testing';
import { UserService } from './user.service';
import { DBModule2 } from '../db.module';
import { UserModule } from './user.module';

describe('UserService', () => {
  let service: UserService;
  let module: TestingModule;
  beforeAll(async () => {
    const module = await Test.createTestingModule({
      imports: [DBModule2, UserModule],
    }).compile();
    service = module.get<UserService>(UserService);
    await module.init();
  });
  afterAll(async () => {
    await module.close();
  });
  it('should be defined', async () => {
    expect(service).toBeDefined();
    await service.onModuleInit();
  });
  it('onModuleInit', async () => {
    expect(await service.onModuleInit()).toMatchSnapshot();
  });
  it('Create User', async () => {
    expect(await service.setUser(undefined, 'Test', 'Pass')).toMatchSnapshot();
  });
  it('Update User', async () => {
    const result = await service.setUser(undefined, 'Test2', 'Pass2');
    expect(
      await service.setUser(result?.id, 'Test2-2', 'abc', { test: 123 }),
    ).toMatchSnapshot();
  });
  it('Delete User', async () => {
    expect(await service.deleteUser(2)).toMatchSnapshot();
  });
  it('List Users', async () => {
    expect(await service.users()).toMatchSnapshot();
  });
  it('List Users Params', async () => {
    expect(await service.users(['id', 'name'])).toMatchSnapshot();
  });
});
