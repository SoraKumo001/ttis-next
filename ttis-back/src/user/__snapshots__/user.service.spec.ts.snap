// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserService Create User 1`] = `
Object {
  "enable": true,
  "id": 2,
  "info": "{}",
  "name": "Test",
  "password": "ebdf8cc00bc4d9ceee633c56c63b49955769a92ca060825c9b08e4af61326e2b",
}
`;

exports[`UserService Delete User 1`] = `true`;

exports[`UserService List Users 1`] = `
Array [
  User {
    "enable": true,
    "id": 1,
    "info": "{}",
    "name": "admin",
    "password": "e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855",
  },
  User {
    "enable": true,
    "id": 3,
    "info": "{}",
    "name": "Test2-2",
    "password": "ba7816bf8f01cfea414140de5dae2223b00361a396177a9cb410ff61f20015ad",
  },
]
`;

exports[`UserService List Users Params 1`] = `
Array [
  User {
    "id": 1,
    "name": "admin",
  },
  User {
    "id": 3,
    "name": "Test2-2",
  },
]
`;

exports[`UserService Update User 1`] = `
User {
  "enable": true,
  "id": 3,
  "info": "{}",
  "name": "Test2-2",
  "password": "ba7816bf8f01cfea414140de5dae2223b00361a396177a9cb410ff61f20015ad",
}
`;

exports[`UserService onModuleInit 1`] = `undefined`;
